flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -144,12 +145,340 @@ mlir::Value calculateTripCount(fir::FirOpBuilder &builder, mlir::Location loc,
+      unorderedLoops.push_back(nestedLoop);
+
+  if (unorderedLoops.empty())
+    return mlir::success();
+
+  if (unorderedLoops.size() > 1)
+    return mlir::failure();
+
+  fir::DoLoopOp nestedUnorderedLoop = unorderedLoops.front();
+
+  std::array<mlir::Operation *, 3> nestedUnorderedControlOps;
+  nestedUnorderedControlOps[0] =
+      nestedUnorderedLoop.getLowerBound().getDefiningOp();
+  nestedUnorderedControlOps[1] =
+      nestedUnorderedLoop.getUpperBound().getDefiningOp();
+  nestedUnorderedControlOps[2] = nestedUnorderedLoop.getStep().getDefiningOp();
+
+  if (llvm::any_of(nestedUnorderedControlOps,
+                   [](auto *op) { return op == nullptr; }))
+    return mlir::failure();
+
+  llvm::SmallVector<mlir::Value> nestedLiveIns;
+  collectLoopLiveIns(nestedUnorderedLoop, nestedLiveIns);
+
+  llvm::DenseSet<mlir::Value> outerLiveInsSet;
+  llvm::DenseSet<mlir::Value> nestedLiveInsSet;
+
+  // Returns a "unified" view of an mlir::Value. This utility checks if the
+  // value is defined by an op, and if so, return the first value defined by
+  // that op (if there are many), otherwise just returns the value.
+  //
+  // This serves the purpose that if, for example, `%op_res#0` is used in the
+  // outer loop and `%op_res#1` is used in the nested loop (or vice versa), that
+  // we detect both as the same value. If we did not do so, we might falesely
+  // detect that the 2 loops are not perfectly nested since they use "different"
+  // sets of values.
+  auto getUnifiedLiveInView = [](mlir::Value liveIn) {
+    return liveIn.getDefiningOp() != nullptr
+               ? liveIn.getDefiningOp()->getResult(0)
+               : liveIn;
+  };
+
+  // Re-package both lists of live-ins into sets so that we can use set equality
+  // to compare the values used in the outerloop vs. the nestd one.
+
+  for (auto liveIn : nestedLiveIns)
+    nestedLiveInsSet.insert(getUnifiedLiveInView(liveIn));
+
+  mlir::Value outerLoopIV;
+  for (auto liveIn : outerLoopLiveIns) {
+    outerLiveInsSet.insert(getUnifiedLiveInView(liveIn));
+
+    // Keep track of the IV of the outerloop. See `isPerfectlyNested` for more
+    // info on the reason.
+    if (outerLoopIV == nullptr)
+      outerLoopIV = getUnifiedLiveInView(liveIn);
+  }
+
+  // For the 2 loops to be perfectly nested, either:
+  // * both would have exactly the same set of live-in values or,
+  // * the outer loop would have exactly 1 extra live-in value: the outer
+  //   loop's induction variable; this happens when the outer loop's IV is
+  //   *not* referenced in the nested loop.
+  bool isPerfectlyNested = [&]() {
+    if (outerLiveInsSet == nestedLiveInsSet)
+      return true;
+
+    if ((outerLiveInsSet.size() == nestedLiveIns.size() + 1) &&
+        !nestedLiveInsSet.contains(outerLoopIV))
+      return true;
+
+    return false;
+  }();
+
+  if (!isPerfectlyNested)
+    return mlir::failure();
+
+  return collectLoopNest(nestedUnorderedLoop, nestedLiveIns, loopNest);
```

Comment:
Note that recursion is [discouraged in MLIR](https://mlir.llvm.org/getting_started/DeveloperGuide/#style-guide). However, recursion into nested regions is "tolerated".

Reply:
Converted to an iterative version.